### DISCUSSION TITLE
Setting vehicle routes with origin/destination or matrixMappers.

### DISCUSSION THREAD

**Discussion by NoNoid** [eclipse/mosaic#110] [Q&A]
Hello i am trying to set Vehicle Routes with the 'matrixMappers' field of the mapping config or the 'origin' or 'destination' fields of the vehicle class. MOSAIC accepts the the example below (with actual coordinates) and runs the simulation but no vehicles are spawned. What am i missing here?

```json
{
  "prototypes": [
    {
      "name": "Car",
      "length": 4.0,
      "minGap": 2.5,
      "maxSpeed": 70.0,
      "vehicleClass": "Car",
      "accel": 2.6,
      "decel": 4.5,
      "sigma": 0.5,
      "tau": 1.0,
      "speedFactor": 1.0,
      "deviations": {
        "length": 0.0,
        "minGap": 0.0,
        "maxSpeed": 0.0,
        "speedFactor": 0.1,
        "accel": 0.0,
        "decel": 0.0,
        "tau": 0.0
      }
    }
  ],
  "vehicles": [
    {
      "startingTime": 5.0,
      "maxTime": 700.0,
      "targetFlow": 1800.0,
      "spawningMode": "CONSTANT",
      "maxNumberVehicles": 500,
      "laneSelectionMode": "BEST",
      "departSpeed": 0.0,
      "departSpeedMode": "MAXIMUM",
      "types": [
        {
          "name": "Car",
          "weight": 1.0
        }
      ],
	  "origin": {
		"center": {
		    "latitude": ..., 
			"longitude": ...
		},
		"radius": 3000
	  },
	  "destination": {
		"center": {
			"latitude": ...,
			"longitude": ...
		},
		"radius": 3000
	  },
      "deterministic": true
    }
  ],
  "matrixMappers":[
	{
		"points": [
			{
				"name": "A",
				"position": {
					"center": {
						"latitude": ..., 
						"longitude": ...
					},
					"radius": 3000
				}
			},
			{
				"name": "B",
				"position": {
					"center": {
						"latitude": ...,
						"longitude": ...
					},
					"radius": 3000
				}
			}
		],
		"types": [
			{
			  "name": "Car",
			  "weight": 1.0
			}
		],
		"odValues": [
			[1000, 2000],
			[3000, 4000]
		]
	}
  ]
}
```

**Comment by schwepmo** [comment]
From first glance your configuration looks correct. Notice however that you don't have to specify `vehicles` when using `matrixMappers`, so you're actually defining two separate flows.

I'm currently investigating if there are further internal issues with `matrixMappers` and will update you as soon as I have results.

Update: There are indeed some additional issues internally which will probably fixed within a week

**Reply by NoNoid** [reply]
I included both `vehicles` and `matrixMappers` because neither worked for me.
The only thing that works for me, is using the routes that i declared during secenario conversion from OSM (`--generate_routes`). Maybe i am missing some steps during scenario conversion?

**Reply by schwepmo** [reply]
Oh I will have a look toady if there are further issues with defining `origin` and `destination`. We rarely ever use these methods of defining vehicle flows but instead generate routes prior and then assign them as you mentioned. Though the `origin` and `destination` are good for prototyping.

**Reply by schwepmo** [reply]
Alright I just checked, both methods should work again once [this](https://github.com/eclipse/mosaic/pull/120) PR is merged. Sorry for that inconvenience.

**Reply by kschrab** [reply]
I merged the mentioned PR (#120). Once you've cloned and build MOSAIC from sources, you'll be able to use origin/destination or matrixMappers

**Reply by fabio-r-goncalves** [reply]
Sorry, I know that this subject already has some time but, can you tell me if this is already fixed in version 22.1 that is available for download on the https://www.dcaiti.tu-berlin.de/ website? I am having the same problem. 
Thx in advance.

**Reply by schwepmo** [reply]
Hey, the fix should indeed be included in version 22.1.  And I just checked the functionality by adding an arbitrary `matrixMapper` to the Barnim scenario. 
Have you looked at the log files generated by MOSAIC? Especially, the `Mapping.log` file?
